core/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -139,14 +140,39 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
         final AtomicArray<BulkItemResponse> responses = new AtomicArray<>(bulkRequest.requests.size());
 
         if (needToCheck()) {
-            // Keep track of all unique indices and all unique types per index for the create index requests:
+            // Attempt to create all the indices that we're going to need during the bulk before we start.
+            // Step 1: collect all the indices in the request
             final Set<String> autoCreateIndices = bulkRequest.requests.stream()
                 .map(DocWriteRequest::index)
                 .collect(Collectors.toSet());
-            final AtomicInteger counter = new AtomicInteger(autoCreateIndices.size());
+            /* Step 2: filter that to indices that don't exist and we can create. At the same time build a map of indices we can't create
+             * that we'll use when we try to run the requests. */
+            Map<String, IndexNotFoundException> indicesThatCannotBeCreated = null;
             ClusterState state = clusterService.state();
-            for (String index : autoCreateIndices) {
-                if (shouldAutoCreate(index, state)) {
+            for (Iterator<String> itr = autoCreateIndices.iterator(); itr.hasNext();) {
```

Comment:
I see. I think it will die young so it's probably not a big deal. I'm fine if you think it's important yet I think either way the rest of my suggestion (to avoid the iterate/remove) stand?